### DISCUSSION TITLE
Imperative mapping of a property to a table column

### DISCUSSION THREAD

**Discussion by petarlishov** [sqlalchemy/sqlalchemy#9860] [Usage Questions] [answered]
Hi, I have a dataclass that looks like this:
```python
@dataclass
class Component:
    id: int = field(init=False)
    service: str
    package: str
    package_version: str
    file: str

    @property
    def external_id(self):
        return f"{self.service}/{self.package}/{self.package_version}{self.file}"
```

I am trying to map the `Component.external_id` property value to an `external_id` column in the `component` table in the database by using imperative mapping, but when I do so, it always sets the external_id in the database to `null`...

```python
mapper_registry = registry()

component_table = Table(
    "component",
    mapper_registry.metadata,
    Column("id", Integer, primary_key=True),
    Column("service", String(255)),
    Column("package", String(255)),
    Column("package_version", String(255)),
    Column("file", String(4096)),
    Column("external_id", String(5000))
)


def start_mapper():
    """Map objects to their corresponding database table."""
    mapper_registry.map_imperatively(Component, component_table)
```

I have seen there are some slight references to how one can do that if they do hybrid or declarative mapping, but I am looking for a clear separation of the persistence layer through a repository and a domain layer that does not depend on the persistence, thus trying to not merge the two into one. The reason I want to be able to persist this column is that we will need to query on it and it is not going to be easy to reliably parse the individual fields out from the external_id (we will be provided with the external_id, the id that is the primary key will not be exposed)

**Comment by zzzeek** [comment]
the only thing getting in the way is the `@property` you have there, which doesn't store any value, so you'd just want to move that out of the way.  it's not clear exactly what the purpose of this property is if you are seeking to persist that value inside the `external_id` column itself, unless you'd like to persist `external_id` under a different name perhaps.

**Reply by petarlishov** [reply]
Yeah, I was looking at having the external_id table column be populated by the value returned by the `Component.external_id` property (the one annotated with `@property`). It does not have to be a property, but looking to have it be exposed as something on the `Component` class and still auto-generated in the way this external_id property does. If I can map the output from a method to a table column, then that would do as well. Are you saying that simply removing the `@property` and making the Component.external_id into a method do the job?

**Reply by zzzeek** [reply]
if you put a method on your class called `def external_id()`, and then whether or not you make a `@property` out of it, if you then do this:

```py
my_object = Component(....)
session.add(my_object)
session.commit()
```

nothing called your `external_id` function.   So the code you put inside that function that creates that string value wasn't called and there's nothing to store.

from the information given here, so far it sounds like you are looking for something like this, here using the dataclass [`__post_init__`](https://docs.python.org/3/library/dataclasses.html#post-init-processing) feature:


```py
@dataclasses.dataclass
class Component:
    id: int = field(init=False)
    service: str
    package: str
    package_version: str
    file: str
    external_id: str = dataclasses.Field(init=False)

    def __post_init__(self):
        self.external_id = f"{self.service}/{self.package}/{self.package_version}{self.file}"

```

**Reply by petarlishov** [reply]
Ooh, that is a cool feature I did not know about! I think that could be helpful!

I like the idea of a method/property mapping to a database column because even if you change one of the other fields after init, the value coming out of the property/method would always be up to date when it comes time to persist it.

E.g. if I initialised a component (for example I fetched it from the DB), edited the component name and then persisted it again after, the external_id field would no longer be correct unless if I also made sure to proactively re-calculate it. I would need to take care to always manually update the external_id when another field is updated.

If there is no way to map a method/property to a column, then I'm happy to do this manually, it should be fine

**Reply by zzzeek** [reply, marked as answer]
> I like the idea of a method/property mapping to a database column because even if you change one of the other fields after init, the value coming out of the property/method would always be up to date when it comes time to persist it.

oh OK that makes it more clear what you actually wanted this to do.    

we support this kind of thing using events, which could be done with a bit of intricacy as the object is mutated in Python using either [@validates](https://docs.sqlalchemy.org/en/20/orm/mapped_attributes.html#sqlalchemy.orm.validates) or [AttributeEvents](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.AttributeEvents),  or using a persist level event, in this case [before_insert](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.MapperEvents.before_insert) and before_update, something like this:


```py

from sqlalchemy import event

@event.listens_for(Component, "before_insert")
@event.listens_for(Component, "before_update")
def _set_external_id(mapper, connection, target):
    target.external_id = f"{target.service}/{target.package}/{target.package_version}{target.file}"


```

there is also the concept of a database "computed" field that would do this, if you were using supporting databases, where the database itself would deliver this string for a particular column.     The feature here seems to be a "client side computed" field, we do have column-level default generators that could also do this, but the default generation function (documented [here](https://docs.sqlalchemy.org/en/20/core/defaults.html)) is a Core-level construct and it doesn't have access to the ORM object, just the row parameters being inserted, which could be used but would not be as convenient.   So the feature would be an ORM level client-side computed/default type of thing.
